### PR TITLE
MultipleProvisionerCsiSnapshotDeleteBackupAndRestore testcase is to test restore of namespaces with multiple provisioners using backup when snapshot is deleted

### PR DESCRIPTION
**Pull request by sn-px**
This MultipleProvisionerCsiSnapshotDeleteBackupAndRestore testcase to test restore of namespaces with multiple provisioners using backup when snapshot is deleted

Ignore reviewing the backup helper changes in this PR (will rebase it once https://github.com/portworx/torpedo/pull/2151 is merged)
vendored in IBM modules to delete csi snapshot from vpc service

**Which issue(s) this PR fixes** (optional)
Closes #https://portworx.atlassian.net/browse/PB-5976

**Special notes for your reviewer**:
https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/Cloud_BYOC/191/execution/node/179/log/

Latest log : https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/Cloud_BYOC/208/execution/node/179/log/

### CODE REVIEW

File: tests/backup/backup_multiple_provisioner_test.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+
+			err = CreateApplicationClusters(BackupOrgID, "", "", ctx)
+			dash.VerifyFatal(err, nil, "Creating source and destination cluster")
+
+			clusterStatus, err := Inst().Backup.GetClusterStatus(BackupOrgID, SourceClusterName, ctx)
+			log.FailOnError(err, fmt.Sprintf("Fetching [%s] cluster status", SourceClusterName))
+			dash.VerifyFatal(clusterStatus, api.ClusterInfo_StatusInfo_Online, fmt.Sprintf("Verifying if [%s] cluster is online", SourceClusterName))
+
+			//sourceClusterUid, err = Inst().Backup.GetClusterUID(ctx, BackupOrgID, SourceClusterName)
```
Comment (kshithijiyer-px) [inline tests/backup/backup_multiple_provisioner_test.go:100]: Do you really need this and the following line?

File: tests/backup_helper.go [hunk 2]
```diff
@@ -353,6 +353,118 @@ var (
...
+		},
+	},
+}
+
+// GetProvisionerDefaultSnapshotMap returns a map with provisioner to default volumeSnapshotClass mappings for the specified cloud provider
+func GetProvisionerDefaultSnapshotMap(cloudProvider string) map[string]string {
+	provisionerSnapshotMap := make(map[string]string)
+	provisionerMap, ok := CloudProviderProvisionerSnapshotMap[cloudProvider]
+	if !ok {
+		//return provisionerSnapshotMap, nil
```
Comment (kshithijiyer-px) [inline tests/backup_helper.go:408]: Remove this if not needed

**Review by kshithijiyer-px — COMMENTED** [review]
A few minor comments

File: tests/backup_helper.go [hunk 3]
```diff
@@ -353,6 +353,118 @@ var (
...
+		},
+	},
+	"ibm": {
+		"vpc.block.csi.ibm.io": {
+			snapshotClasses: []string{},
+			defaultSnapshot: "ibmc-vpcblock-snapshot",
+			appList:         []string{"postgres-csi"},
+		},
+	},
+	"aws": {
```
Comment (ak-px) [inline tests/backup_helper.go:376]: Please add TODO string, it its placeholder for snapshotClasses

File: tests/backup/backup_multiple_provisioner_test.go [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+		ctx, err := backup.GetAdminCtxFromSecret()
+		log.FailOnError(err, "Fetching px-central-admin ctx")
+		opts := make(map[string]bool)
+		opts[SkipClusterScopedObjects] = true
+
+		log.InfoD("Deleting the deployed apps after the testcase")
+		DestroyApps(allAppContext, opts)
+
+		log.InfoD("switching to default context")
```
Comment (ak-px) [inline tests/backup/backup_multiple_provisioner_test.go:186]: Do we need this.

File: tests/common.go [hunk 5]
```diff
@@ -5636,6 +5667,76 @@ func DeleteBucket(provider string, bucketName string) {
...
+	if GetClusterProvider() == "ibm" {
+		err = DeleteIbmSnapshotsForVolumes(volumes)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func DeleteIbmSnapshotsForVolumes(volumeNames []string) error {
```
[the remark this answers is not in the document]
Comment (sn-px) [reply tests/common.go:5685]: done

File: tests/common.go [hunk 6]
```diff
@@ -5636,6 +5667,76 @@ func DeleteBucket(provider string, bucketName string) {
 	})
 }
 
+// DeleteSnapshotsForVolumes for all the volumes from the
```
[the remark this answers is not in the document]
Comment (sn-px) [reply tests/common.go]: done

File: tests/backup_helper.go (hunk 2 above)
Comment (sn-px) [reply tests/backup_helper.go]: done

File: tests/backup_helper.go [hunk 7]
```diff
@@ -353,6 +353,118 @@ var (
...
+// GetProvisionerDefaultSnapshotMap returns a map with provisioner to default volumeSnapshotClass mappings for the specified cloud provider
+func GetProvisionerDefaultSnapshotMap(cloudProvider string) map[string]string {
+	provisionerSnapshotMap := make(map[string]string)
+	provisionerMap, ok := CloudProviderProvisionerSnapshotMap[cloudProvider]
+	if !ok {
+		return provisionerSnapshotMap
+	}
+
+	for provisioner, info := range provisionerMap {
+		println(info.defaultSnapshot)
```
Comment (kshithijiyer-px) [inline tests/backup_helper.go:413]: Can you remove it if you have added for debugging?

**Review by kshithijiyer-px — APPROVED** [review]
Minor comment looks like you have forgotten to remove debug statements rest looks good to me.

**Review by ak-px — APPROVED** [review]
LGTM

File: tests/backup_helper.go (hunk 7 above)
Comment (sn-px) [reply tests/backup_helper.go]: Done

File: tests/common.go [hunk 8]
```diff
@@ -5636,6 +5667,79 @@ func DeleteBucket(provider string, bucketName string) {
 	})
 }
 
+// DeleteSnapshotsForVolumes deletes snapshots for the specified volumes based on cluster provider
+func DeleteSnapshotsForVolumes(volumes []string) error {
```
Comment (mkoppal-px) [inline tests/common.go:5671]: Shouldn't this be part of volume driver?

File: tests/backup/backup_multiple_provisioner_test.go [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+	)
+
+	JustBeforeEach(func() {
+		StartPxBackupTorpedoTest("MultipleProvisionerCsiSnapshotDeleteBackupAndRestore", "Delete Csi snapshot and restore namespaces from backup", nil, 296725, Sn, Q4FY24)
+
+		backupLocationMap = make(map[string]string)
+		providers = GetBackupProviders()
+
+		// Deploy application for Default backup
+		appSpecList = []string{"postgres"}
```
Comment (mkoppal-px) [inline tests/backup/backup_multiple_provisioner_test.go:52]: Why is this hard-coded here?

Comment (sn-px) [reply tests/backup/backup_multiple_provisioner_test.go:52]: Reading from the current applist may not work , will need to impl getting only the prefix will discuss with you

File: tests/common.go (hunk 8 above)
Comment (sn-px) [reply tests/common.go:5668]: Done. moved to volume driver

**Review by mkoppal-px — APPROVED** [review]
LGTM

**Pull request merged**